Lab08 - Wi-Fi (#108)
UPB-FILS-MA/upb-fils-ma.github.io: docs/lab/08/index.md

---
sidebar_position: 8
description: Wi-Fi
slug: /lab/08
---

# 08 - Wi-Fi

This lab will teach you the basics of networking and Wi-Fi, and how to use Wi-Fi with the Raspberry Pi Pico W, in Embassy.

## Resources

TBD

## Networking basics

A network is a set of devices that are linked together in order to share resources. Devices can talk to one another using different communication protocols, usually Ethernet (wired) or Wi-Fi (wireless).

Devices that communicate in a network have an associated IP (Internet Protocol) address. There are two types of IP addresses: IPv4 (stored on 32 bits) and IPv6 (newer, stored on 64 bits). We will be discussing only the IPv4 in this lab, since it is most commonly used in local networks.

### IPv4

An IPv4 address looks like this:

> `192.168.0.1/8`

Reviewer: ```suggestion
`192.168.0.1/24`
```

> Each number separated by a dot is represented on 8 bits (ranging from 0 to 255). The number after the slash is the *mask* of the address, which will be described shortly.

Reviewer: ```suggestion
Each number separated by a dot is represented on 8 bits (ranging from 0 to 255). The number after the slash is the *network mask*, which will be described shortly.
```

An IP address is made of two parts:
- network prefix: defines the network the device is a part of
- host identifier: is unique to the device in the network

> The *mask* of an address defines the number of bits that are used to represent the host identifier. In other words, the bigger the mask, the more unique devices can exist in the network.

Reviewer: ```suggestion
The *network mask* of an address defines the number of bits that are used to represent the network prefix. In other words, the smaller the mask, the more unique devices can exist in the network.
```

Author: Done.

> For example, for the above address `192.168.0.1/8`, we know that the host identifier is 8 bits long. Therefore, the first 24 bits give us the network prefix and the last 8 bits give us the host identifier.

Reviewer: ```suggestion
For example, for the above address `192.168.0.1/24`, we know that the network prefix is 8 bits long. Therefore, the first 24 bits give us the network prefix and the last 8 bits give us the host identifier.
```

Any device that wants to communicate in this local network (or *subnet*) must have an address ranging from `192.168.0.1` to `192.168.0.254`.

:::note

> In the above example, `192.168.0.0` is the *network address*, and `192.168.0.255` is the *broadcast address*. The first and last possible addresses in a subnet are reserved and should not be used for devices.

Author: Done. I tried to avoid delving too deep into this part, since it should be part of a networking subject.

- 5 GHz radio band, super high frequency
- others, not widely used

Each radio frequency range is divided into a multitude of *channels*. Each device that uses a frequency band communicates over one channel. Ideally, each device is allocated a separate channel. Sometimes, though, when there are many devices using a Wi-Fi frequency band, devices might end up having to *share* a channel and wait their turn to transmit data, leading to lower communication speeds. 5 GHz provides more channels than 2.4 GHz, allowing more devices to use the band without interference. 

### Wi-Fi on the Raspberry Pi Pico

Apart from the RP2040 microcontroller, the Raspberry Pi Pico board also has a separate, on-board wireless interface. It is a 2.4 GHz Infineon chip, *CYW43439*. This chip deals only with Wi-Fi (and Bluetooth) communication, and it is connected to the RP2040 via SPI.

![cyw43](images/cyw43.png)

### Security

#### Open Wi-Fi

Open wireless networks don't require any authentication, and all transmissions are unencrypted. Any device that connects to an open network can easily intercept data sent in this network. This is unsecured and unsafe.

#### Wired Equivalent Privacy (WEP)

> WEP is the first security protocol ever implemented. It uses data encryption based on generated key values. It was designed in 1197 and has since become obsolete, because it is really easy to crack and basically as unsafe as open Wi-Fi.

Reviewer 2: ```suggestion
WEP is the first security protocol ever implemented. It uses data encryption based on generated key values. It was designed in 1997 and has since become obsolete, because it is really easy to crack and basically as unsafe as open Wi-Fi.
```
No?

Author: Whoops

    info!("rxd {}", from_utf8(&buf[..n]).unwrap());

    match socket.write_all(&buf[..n]).await {
        Ok(()) => {}
        Err(e) => {
            warn!("write error: {:?}", e);
            break;
        }
    };
}
```

## Exercises

1. The lab skeleton contains an example of setting up the Wi-Fi and connecting to an access point.
- Scan for local networks. (**1p**)
- Modify the code in order to connect to an existing network. Once connected, determine the IP address that has been assigned to the Pico through DHCP. (**1p**)
- Use a static IP address instead of a dynamic one. You can use the IP determined at the previous point. (**1p**)

2. Test the example by connecting your computer to the Pico through Wi-Fi. Use `netcat` or any alternative to send a TCP packet to the Pico server. You should receive back the same string. (**1p**)

:::tip
On Windows, you can use [NCat](https://nmap.org/download.html#windows). The command for establishing a TCP connection is:
```shell
ncat [ip_address] [port]
```
To send a TCP packet, write any string in the terminal and press enter. You should receive back the same string, since this is what our server is currently configured to do.
:::

> 3. Modify the way the server on the Pico deals with connections. Send a message through Wi-Fi to your computer whenever a button is pressed. (**2p**)

Reviewer 2: I would use UDP for this.

---
sidebar_position: 8
description: Wi-Fi
slug: /lab/08
---

# 08 - Wi-Fi

This lab will teach you the basics of networking and Wi-Fi, and how to use Wi-Fi with the Raspberry Pi Pico W, in Embassy.

## Resources

**Andrew Tanenbaum**, *Computer networks (5th edition)*
  - Chapter 1 - *Introduction*
    - Subchapter 1.4.2 - *The TCP/IP Reference Model*

## Networking basics

> A network is a set of devices that are linked together in order to share resources. Devices can talk to one another using different communication protocols, usually Ethernet (wired) or Wi-Fi (wireless).

Reviewer 2: Protocols are at different levels, Ethernet and WiFi are L2.

> Devices that communicate in a network have an associated IP (Internet Protocol) address. There are two types of IP addresses: IPv4 (stored on 32 bits) and IPv6 (newer, stored on 64 bits). We will be discussing only the IPv4 in this lab, since it is most commonly used in local networks.

Reviewer 2: This is for TCP/IP networks only. Of course, these are the most commonly used. In any network devices have addresses, in IP networks these are IP addresses.

### IPv4

An IPv4 address looks like this:
`192.168.0.1/24`
Each number separated by a dot is represented on 8 bits (ranging from 0 to 255). The number after the slash is the *network mask*, which will be described shortly.

An IP address is made of two parts:
- network prefix: defines the network the device is a part of
- host identifier: is unique to the device in the network

The *network mask* defines the number of bits that are used to represent the network prefix. In other words, the smaller the mask, the more unique devices can exist in the network.

For example, for the above address `192.168.0.1/24`, we know that the network prefix is 24 bits long. Therefore, the first 24 bits give us the network prefix and the last 8 bits give us the host identifier. The mask in this case would look like this: `255.255.255.0`.

:::info
In binary, `255.255.255.0` is `11111111.11111111.11111111.00000000`. 

> First 24 bits are 1, last 8 bits are 0. 

Reviewer 2: ```suggestion
The first 24 bits are `1`, the last 8 bits are `0`. 
```

:::

> Any device that wants to communicate in this local network (or *subnet*) must have an address ranging from `192.168.0.1` to `192.168.0.254`.

Reviewer 2: Explain here that the first 24 bits are the same. You can use wavedrom to draw the address and have different colours for the bits.

```
{
    "reg": [
        {
            "bits": 5,
            "name": "Recipient",
            "type": 7,
            "attr": [
                "00000 - Device",
                "00001 - Interface",
                "00010 - Endpoint",
                "00011 - Other",
                "00100 - 11111 - Reserved"
            ]
        },
        {
            "bits": 2,
            "name": "Type",
            "type": 5,
            "attr": [
                "00 - Standard",
                "01 - Class",
                "10 - Vendor",
                "11 - Reserved"
            ]
        },
        {
            "bits": 1,
            "name": "Direction",
            "type": 3,
            "attr": [
                "0 - Host to Device",
                "1 - Device to Host"
            ]
        }
    ],
    "config": {
        "hspace": 950
    }
}

```

:::note
In the above example, `192.168.0.0` is the *network address*, and `192.168.0.255` is the *broadcast address* (used to broadcast to *all* endpoints in a subnet). The first and last possible addresses in a subnet are reserved and should not be assigned to devices.

To determine the *network address*, we need to perform a logic `AND` operation between the IP address and the network mask.

> To determine the *broadcast address*, we need to perform a logic `OR` operation between the IP address and the network mask.

Reviewer 2: I think it is IP or ~Mask (inverse)

This means converting the IP address and network mask to binary, then performing `AND` or `OR` on every bit, then converting it back to decimal.
:::

Learn more about subnetting [here](https://networklessons.com/cisco/ccna-routing-switching-icnd1-100-105/what-is-subnetting).

### Routers

> Routers are devices that forward packets between networks. The packets of the sender will be *routed* to the receiver of a different network, depending on the IP of the destination, passing from one router to another until they reach their destination. This happens through complex routing protocols. Routers also deal with traffic filtering, DNS (Domain Name System), DHCP and others.

Reviewer 2: You can use the image from the slides.

The *default gateway* in a local network is the router where packets must be sent when the *destination IP* is not in the current local network. From there, packets will be routed further. Read more about it [here](https://en.wikipedia.org/wiki/Default_gateway).

### DHCP

> Dynamic Host Configuration Protocol (DHCP) is a network management protocol used by routers to dynamically allocate IP addresses to devices connected to its network. Routers dynamically assign addresses to new devices in the network, if they don't already have a *static* IP.

Reviewer 2: DHCP leases IP, Mask, Gateway and DNS servers.

### Networking protocols

#### TCP (Transmission Control Protocol)

> TCP is *connection-oriented*, meaning that a connection must be first established between two devices (client and server) before data can be sent. Packets that are not received correctly must be *retransmitted* by the sender, ensuring that data always reaches its destination. TCP is used in cases where data integrity is most important.

Reviewer 2: The packet might be fragmented during the transmission and the receiver might receive it in several smaller packets.

#### UDP (User Datagram Protocol)

> UDP, on the other hand, is *connectionless*, and prioritizes time over reliability. Packets that are not received correctly are dropped and **not** retransmitted. UDP is used in time-sensitive applications, such as audio streaming.

Reviewer 2: The packet is not fragmented.

    }
}
```

## Exercises

1. The lab skeleton contains an example of setting up the Wi-Fi and connecting to an access point.
- Scan for local networks. (**1p**)
- Modify the code in order to connect to an existing network. Once connected, determine the IP address that has been assigned to the Pico through DHCP and the default gateway of the network. (**1p**)
:::tip
You can use `stack.config_v4()` to find the assigned IP address and default gateway. 
:::
- Use a static IP address instead of a dynamic one. You can use the IP determined at the previous point. (**1p**)

2. Test the example by connecting your computer to the Pico through Wi-Fi. Use `netcat` or any alternative to send a TCP packet to the Pico server. You should receive back the same string. (**1p**)

:::tip
On Windows, you can use [NCat](https://nmap.org/download.html#windows). The command for establishing a TCP connection is:
```shell
ncat [ip_address] [port]
```
To send a packet, write any string in the terminal and press enter. You should receive back the same string, since this is what our server is currently configured to do.
You can run `ncat -h` to see a list of other available parameters for this command.
:::

> 3. Modify the way the server on the Pico deals with connections. Send a message through Wi-Fi to your computer whenever a button is pressed. It's better to use UDP for this. (Why?) (**2p**)

Reviewer 2: The Why is a question for us the reviewers? If so:
- the packet os received fully, not fragmented; tcp might fragment the packet
- the local network should not loose a lot of packets

Author: I was debating whether to have the students try and answer the question or not. I removed it in the end.

:::tip
You can transform a string slice to bytes using `as_bytes()`.
:::

> 4. Modify the server so that you can *also* send commands from your computer to turn an LED on and off. (**1p**)

Reviewer 2: I would describe how the sent message should look like.

:::tip
Use *select* to await multiple futures (wait for button press and read from socket). Refer to Lab 5.
:::

5. Configure the Pico as an access point. You will need to set a static IP address both for the Pico and your computer, to make sure they are both in the same subnet, so that they can communicate. (**1p**)

> 6. Connect two Picos together through Wi-Fi. Control the LED connected to one Pico through a button connected to the second Pico. One team can configure the server for the LED Pico, and another team the server for the button Pico. (**2p**)

Reviewer 2: I would describe how the sent message should look like.
